doc/sphinx/proof-engine/ltac.rst

```diff
@@ -783,46 +825,85 @@ local definitions) with:
+   The pattern :n:`_` matches any term and stops further evaluation of match patterns.
+
+   **no pattern variables in the form ?1 as in the Delahaye paper?**
+   **link to paper, mention that some things have changed, e.g tactic are lowercase**
+
+   Patterns may contain pattern variables in the form :n:`?@ident`.
+   If a pattern variable occurs more than once in a pattern, it must have
+   the same value for each occurrence for it to match.  Pattern variables are
+   substituted into each :n:`ltac_expr__i` before it's applied.
+
+   Matching is first-order except for variables in the form :n:`@?id`
+   that occur in the head position of an application. For these variables,
+   matching is second-order and returns a functional term.  **huh?  example??**
 
    Alternatively, when a metavariable of the form :n:`?id` occurs under binders,
    say :n:`x__1, …, x__n` and the expression matches, the
    metavariable is instantiated by a term which can then be used in any
    context which also binds the variables :n:`x__1, …, x__n` with
    same types. This provides with a primitive form of matching under
-   context which does not require manipulating a functional term.
+   context which does not require manipulating a functional term.  **huh?**
+
+   **I'd like to drop the next paragraph.  Not sure if it says something important**
+   **"is in position to be applied to a goal"??**
 
-   If the matching with :n:`@cpattern__1` succeeds, then :n:`@ltac_expr__1` is
-   evaluated into some value by substituting the pattern matching
-   instantiations to the metavariables. If :n:`@ltac_expr__1` evaluates to a
+   If :n:`@ltac_expr__i` evaluates to a
    tactic and the match expression is in position to be applied to a goal
    (e.g. it is not bound to a variable by a :n:`let in`), then this tactic is
    applied. If the tactic succeeds, the list of resulting subgoals is the
-   result of the match expression. If :n:`@ltac_expr__1` does not evaluate to a
+   result of the match expression. If :n:`@ltac_expr__i` does not evaluate to a
    tactic or if the match expression is not in position to be applied to a
-   goal, then the result of the evaluation of :n:`@ltac_expr__1` is the result
+   goal, then the result of the evaluation of :n:`@ltac_expr__i` is the result
    of the match expression.
 
-   If the matching with :n:`@cpattern__1` fails, or if it succeeds but the
-   evaluation of :n:`@ltac_expr__1` fails, or if the evaluation of
-   :n:`@ltac_expr__1` succeeds but returns a tactic in execution position whose
-   execution fails, then :n:`cpattern__2` is used and so on. The pattern
-   :n:`_` matches any term and shadows all remaining patterns if any. If all
-   clauses fail (in particular, there is no pattern :n:`_`) then a
-   no-matching-clause error is raised.
+   **The following description of the match_keys is probably wrong.  Théo, can you give**
+   **me examples that show the different behavior?**
```

Comment:
All I can give you is this example:

```
Goal True.
match False with
| False => idtac "Foo"; fail
| _ => easy
end. (* succeeds after printing "Foo" *)
Abort.

Goal True.
lazymatch False with
| False => idtac "Foo"; fail
| _ => easy
end. (* fails after printing "Foo" *)
```

Reply:
What am I doing wrong here?  This:

```
      .. coqtop:: reset all warn fail

         Goal True.
         match False with
         | False => idtac "try"; fail
         | _ => idtac "success"
         end.
```

produces

```
Extension error:
/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq/doc/sphinx/proof-engine/ltac.rst:: Error while sending the following to coqtop:
    lazymatch False with | False => idtac "try"; fail | _ => idtac "success" end.
  coqtop output:
    try
    Toplevel input, characters 0-77:
    > lazymatch False with | False => idtac "try"; fail | _ => idtac "success" end.
    > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    Error: Tactic failure.

  Full error text:
    End Of File (EOF). Exception style platform.
```

Reply:
Ah, I put the "warn fail" on the wrong line.